runtime/src/accounts_db.rs

```diff
@@ -2200,33 +2212,156 @@ impl AccountsDb {
-            // `lock` released here
+        // If there are no entries in the index, return
+        let (slot, store_id, offset) = self.accounts_index_get(ancestors, pubkey, max_root)?;
+
+        #[cfg(test)]
+        {
+            // Give some time for cache flushing to occur here for unit tests
+            sleep(Duration::from_millis(self.load_delay));
+        }
+
+        let mut account_accessor =
+            self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset);
+
+        // Failsafe for potential race conditions with cache flush and accounts clean
+        let mut num_acceptable_failed_iterations = 0;
+        loop {
+            if num_acceptable_failed_iterations >= 10 {
+                // The latest version of the account existed in the index, but could not be
+                // fetched from storage. This means a race occurred between this function and clean
+                // accounts/purge_slots
+                datapoint_warn!(
+                    "accounts_db-do_load_warn",
+                    (
+                        "warn",
+                        format!("do_load() failed to get key: {} from storage, latest attempt was for slot: {}, storage_entry: {}",
+                        pubkey,
+                        slot,
+                        store_id),
+                        String
+                    ),
+                );
+                return None;
+            }
+            match account_accessor {
+                LoadedAccountAccessor::Stored(None) => {
+                    if is_transaction_load {
+                        panic!(
+                            "It should not happen that the storage entry doesn't exist if the entry in
+                            the accounts index is the latest version of this account. This is because clean
+                            should not be removing storage entries that have alive accounts"
+                        );
+                    }
+                    // RPC get_account() may have fetched an old root from the index that was
+                    // either:
+                    // 1) Cleaned up by clean_accounts(), so the accounts index has been updated
+                    // and the storage entries have been removed.
+                    // 2) Dropped by purge_slots() because the slot was on a minor fork, which
+                    // removes the slots' storage entries but doesn't purge from the accounts index
+                    // (account index cleanup is left to clean for stored slots).
+                    else {
+                        // Increment failsafe
+                        num_acceptable_failed_iterations += 1;
+                    }
+                }
+                LoadedAccountAccessor::Cached(None) => {
+                    // Cache was flushed in between checking the index and retrieving from the cache,
+                    // so retry. Don't increment `num_failed_iterations` here because it's impossible
```

Comment:
`num_acceptable_failed_iterations` looks to be incremented; maybe stale comment? (also, variable name is out of sync?)

Reply:
ah yeah updated!